docs/sql-ref-datatypes.md

```diff
@@ -314,3 +314,33 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
 |        3| Infinity|
 +---------+---------+
 ```
+
+#### Data type compatibility
+
+The following is the hierarchy of data type compatibility. In an operation involving different and compatible data types, these will be promoted to the lowest common top type to perform the operation.
+
+For example, if you have an add operation between an integer and a float, the integer will be treated as a float, the least common compatible type, resulting the operation in a float.
+
+The most common operations where this hierarchy is applied are:
```

Comment:
I believe we have much more complicated rules than here ..  see `TypeCoercion.scala`.

Reply:
Yes, I was trying to simplify and put in this list the most popular "operations" but I can extend it. I'm going to review it. Thanks!

Reply:
I'm not sure if explain with a lot of detail this cases is useful. Possibly it's better to remove this list to avoid misunderstand. @HyukjinKwon what do you think?

Reply:
@gatorsmile WDYT? Do you want to have a table to list the most common rules like the one in https://www.ibm.com/support/knowledgecenter/SSEPGG_10.1.0/com.ibm.db2.luw.sql.ref.doc/doc/r0008477.html?

Reply:
Have you checked the other docs, e.g., Oracle, SQL server, ...? If they've already described  type coercsion, I think we are able to refer to them.

Reply:
I have checked Oracle and SQL server documentation.
SQL server don't detail all implicit conversions:
https://docs.microsoft.com/en-us/sql/t-sql/data-types/data-type-conversion-database-engine?view=sql-server-ver15
https://docs.microsoft.com/en-us/sql/t-sql/data-types/data-type-precedence-transact-sql?view=sql-server-ver15
But Oracle have a more detailed rules (see "Implicit Data Conversion")
https://docs.oracle.com/cd/B28359_01/server.111/b28286/sql_elements002.htm#SQLRF00214

Reply:
Ah, I see. IMO the matrix format of the Oracle doc looks better.

Reply:
I guess maybe keep the current Hierarchy compatible types table because it shows the best to worst order, and also add a matrix format data conversion table like Oracle doc? IBM doc actually has two tables, one for Data Type Conversion Precedence List, the other is for casting between data types.

Reply:
I think it's fine to have both tables. I'm going to update the PR to include it